### DISCUSSION TITLE
using-cancellationtokens-in-asp-net-core-minimal-apis/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#10] [Announcements]
# using-cancellationtokens-in-asp-net-core-minimal-apis/

In this post I'll show how you can use CancellationTokens in your minimal API handlers to stop execution when a user cancels a request from their browser

https://andrewlock.net/using-cancellationtokens-in-asp-net-core-minimal-apis/

**Comment by mangeg** [comment]
Could the caveat with catching all ```OperationCanceledException``` exceptions be solved with an exception filter?

```cs
try
{
    await _next( context );
}
catch ( OperationCanceledException ex ) when (ex.CancellationToken == context.RequestAborted)
{
    _logger.LogInformation( "Request was cancelled" );
    context.Response.StatusCode = 409;
}
```

**Comment by RehanSaeed** [comment]
Small correction. 409 is a 'Conflict' response and 499 is actually a non-standard 'Client Closed Request' response that NGINX uses. I actually asked this question on StackOverflow some time ago and the answers were interesting. Only your reverse proxy is going to see the status code, so I think its probably worth doing just to get the logging right and reduce load as you have shown.

https://stackoverflow.com/questions/46234679/what-is-the-correct-http-status-code-for-a-cancelled-request

**Comment by RehanSaeed** [comment]
I've published the above middleware in the [Boxed.AspNetCore](https://www.nuget.org/packages/Boxed.AspNetCore) NuGet package with @mangeg's idea which seems to work. You can see my implementation below.

- https://github.com/Dotnet-Boxed/Framework/blob/main/Source/Boxed.AspNetCore/Middleware/RequestCanceledMiddleware.cs

**Comment by EngincanV** [comment]
Great article as always. Thanks, Andrew.

**Comment by zxbeltm** [comment]
RequestCanceledMiddleware cannot work when url use 0.0.0.0 #507

https://github.com/Dotnet-Boxed/Framework/issues/507

**Comment by StevenTCramer** [comment]
won't 
`app.MapGet("/slowtest", () =>`
and
app.MapGet("/slowtest", (CancellationToken token, ILogger<Handler> logger) =>

require `async`
as in
app.MapGet("/slowtest", async (CancellationToken token, ILogger<Handler> logger) =>

**Reply by andrewlock** [reply]
Yep! Thanks

**Comment by jhogan05** [comment]
Working through you book ASP.Net core in Action, Third edition and keep getting sidetracked by your excellent blog posts.

Just in the example with :

app.MapGet("/slowtest", async (CancellationToken token, ILogger<Handler> logger)

I notice you are still then call “app.Logger” instead of the injected Handler logger.
